bob-access/src/credentials.rs

```diff
@@ -59,64 +73,186 @@ impl CredentialsKind {
     }
 }
 
-impl Credentials {
-    pub fn builder() -> CredentialsBuilder {
-        CredentialsBuilder::default()
-    }
-
-    pub fn ip(&self) -> Option<IpAddr> {
-        Some(self.address?.ip())
-    }
-
-    pub fn is_complete(&self) -> bool {
-        self.address.is_some() && self.ip().is_some()
-    }
-
-    pub fn kind(&self) -> Option<&CredentialsKind> {
-        self.kind.as_ref()
-    }
-}
-
-#[derive(Debug, Default)]
-pub struct CredentialsBuilder {
+#[derive(Default)]
+pub struct RequestCredentialsBuilder {
     kind: Option<CredentialsKind>,
     address: Option<SocketAddr>,
 }
 
-impl CredentialsBuilder {
+impl RequestCredentialsBuilder {
     pub fn with_username_password(
-        &mut self,
+        mut self,
         username: impl Into<String>,
         password: impl Into<String>,
-    ) -> &mut Self {
+    ) -> Self {
         self.kind = Some(CredentialsKind::Basic {
             username: username.into(),
             password: password.into(),
         });
         self
     }
 
-    pub fn with_token(&mut self, token: impl Into<String>) -> &mut Self {
+    pub fn with_token(mut self, token: impl Into<String>) -> Self {
         self.kind = Some(CredentialsKind::Token(token.into()));
         self
     }
 
-    pub fn with_address(&mut self, address: Option<SocketAddr>) -> &mut Self {
+    pub fn with_address(mut self, address: Option<SocketAddr>) -> Self {
         self.address = address;
         self
     }
 
-    pub fn with_nodename(&mut self, node_name: impl Into<String>) -> &mut Self {
+    pub fn with_nodename(mut self, node_name: impl Into<String>) -> Self {
         self.kind = Some(CredentialsKind::InterNode(node_name.into()));
         self
     }
 
-    pub fn build(&self) -> Credentials {
-        Credentials {
+    pub fn build(self) -> RequestCredentials {
+        RequestCredentials {
+            address: self.address,
+            kind: self.kind,
+        }
+    }
+}
+
+const RESOLVE_THRESHOLD: u32 = 3;
+
+#[derive(Debug, Clone)]
+enum ResolveState {
+    Waiting(u32),
+    Pending,
```

Comment:
Names are little bit confusing. I think `Waiting` is better to rename to `Resolved` and `Pending` to `InProgress`